ui/src/app/pages/data/data-set/dataset-page.tsx

```diff
@@ -519,60 +530,73 @@ const DataSetPage = fp.flow(withCurrentWorkspace(), withUrlParams())(
-          case 400:
-            if (exceptionResponse.message ===
-              'Data Sets must include at least one cohort and concept.') {
-              errorText = exceptionResponse.message;
-            } else if (exceptionResponse.message ===
-              'Concept Sets must contain at least one concept') {
-              errorText = 'One or more of your concept sets has no concepts. ' +
-                'Please check your concept sets to ensure all concept sets have concepts.';
-            }
-            break;
-          case 404:
-            if (exceptionResponse.message.startsWith(
-              'Not Found: No Cohort definition matching cohortId')) {
-              errorText = 'Error with one or more cohorts in the data set. ' +
-                'Please submit a bug using the contact support button';
-            }
-            break;
-          case 504:
-            if (exceptionResponse.message ===
-              'Timeout while querying the CDR to pull preview information.') {
-              errorText = 'Query to load data from the All of Us Database timed out. ' +
-                'Please either try again or export data set to a notebook to try there';
-            }
-            break;
-          default:
-            errorText = 'An unexpected error has occurred. ' +
-              'Please submit a bug using the contact support button';
+      const domains = fp.uniq(this.state.selectedValues.map(domainValue => domainValue.domain));
+      this.setState({
+        previewList: domains.map(domain => {
+          return {
+            domain: domain,
+            isLoading: true,
+            values: []
+          };
+        }),
+        previewDataLoading: true,
+        selectedPreviewDomain: domains[0]
+      });
+      domains.forEach(async domain => {
+        const request = {
+          domain: domain,
+          conceptSetIds: this.state.selectedConceptSetIds,
+          includesAllParticipants: this.state.includesAllParticipants,
+          cohortIds: this.state.selectedCohortIds,
+          prePackagedConceptSet: this.getPrePackagedConceptSet(),
+          values: this.state.selectedValues.map(domainValue => domainValue.value)
+        };
+        try {
+          const dataSetPreviewResp = await dataSetApi().generateDataSetPreview(namespace, id, request);
+          const domainPreviewIndex = '[' + fp.findIndex(['domain', domain], this.state.previewList) + ']';
+          this.setState({previewList: fp.set(domainPreviewIndex + '.isLoading', false,
+            fp.set(domainPreviewIndex + '.values', dataSetPreviewResp.values, this.state.previewList))});
+        } catch (ex) {
+          const exceptionResponse = await ex.json() as unknown as ErrorResponse;
+          let errorText: string;
+          switch (exceptionResponse.statusCode) {
+            case 400:
+              if (exceptionResponse.message ===
+                'Data Sets must include at least one cohort and concept.') {
```

Comment:
nit: can all these strings be const

Reply:
Can you explain what you mean? They are currently constants. Do you mean pulling them out to be top level const variables?

I would push back a bit against that, because they are currently used only here for error handling, and I would worry that dong that would add a layer of unnecessary abstraction.

Reply:
This error message (in US English) is NOT to be construed as an inviolable part of the contract. If we need fine-grained validation, let's make a separate return structure that includes that information. We could easily change the wording on the back end without finding this.

Reply:
> Can you explain what you mean? They are currently constants. Do you mean pulling them out to be top level const variables?
> 
> I would push back a bit against that, because they are currently used only here for error handling, and I would worry that dong that would add a layer of unnecessary abstraction.

I meant pulling at the top level it was more for readability.

Reply:
@jaycarlton there's a ticket to build out that structure (with a clear error code that we can parse) That's something that may warrant a bit of a design before doing though.

@NehaBroad I think that pulling it out to the top level would reduce some of the readability here, but am happy to discuss either with faces or more here. Since if we pulled it out it would need to be clear that these are the various preview data error messages, whereas if we keep them here then it is clear via variable placement.

Reply:
https://precisionmedicineinitiative.atlassian.net/secure/RapidBoard.jspa?rapidView=30&projectKey=RW&view=planning&selectedIssue=RW-3006&epics=visible&search=errorcode

@jaycarlton FYI

Reply:
Right, we've chatted about that one. 

In this case, I believe what we need is a separate dataset request validation API that you call before the dataset generation, and which could have detailed error information. Then the main API could just return 400. There are lots of special rules here that don't apply broadly, so I feel like that other ticket is not quite what we need.

Regarding making a constant, I 100% agree we should not have naked string literals in the middle of business logic. It makes translation much harder, for one thing. It also makes the code less reusable. But the real problem here is that you're trying to scan the error text (intended for humans) and take action on it, and that's at the very least a strong code smell. Either the message is good enough for the UI and we just show it verbatim, or we add finer-grained structure to give you enough info to take the right actions. Ultimately, I think driving the UI logic via positive api responses is much more stable; error handling changes frequently (even unintentionally), and it can be more art than science determining which 4xx or 5xx is appropriate.

Reply:
Got it, I am happy to move this to a constant string. Will do.

Trying to scan the error text is 100% the wrong thing. I agree with you completely. I would much prefer us to move to an error code based parsing, as opposed to error text. 

I think I strongly prefer having the API call itself do the validation, as it moves us more into making this API endpoint a reusable piece of code in other places. If another institute decided to build a UI on top of our API (I find this very unlikely, but it is theoretically possible), and used our Data Set Builder, I would hate for them to have to remember to call a separate API endpoint before calling preview. But they should get back a clear and concise immutable value that is part of the API contract, and not error message text that they parse on. That is where I think we should be improving our code, and it contains the logic for validation within the actual endpoint.

Reply:
Talked to Jay in a call. We're punting on this for now, but have a ticket to follow up on error responses throughout our codebase.